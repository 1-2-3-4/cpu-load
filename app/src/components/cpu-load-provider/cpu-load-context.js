import * as React from 'react'

export const CPULoadContext = React.createContext([])