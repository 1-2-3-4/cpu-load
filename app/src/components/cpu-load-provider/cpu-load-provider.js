import * as React from 'react'

import { CPULoadContext } from './cpu-load-context'
import { maintainWindow } from '../../utils/maintain-window'

const TEN_SECONDS = 10000
const TEN_MINUTES = 600000
const WINDOW_LENGTH = TEN_MINUTES / TEN_SECONDS

export const CPULoadProvider = ({ children }) => {
  const [cpuLoad, setCpuLoad] = React.useState([])

  const fetchCPU = React.useCallback(
    async () => {
      const { loadAverage } = await fetch('http://localhost:3000/cpu-load')
        .then(response => response.json()
          .then(data => data)
        )

      setCpuLoad(cpuLoad => maintainWindow([...cpuLoad, (loadAverage * 100)], WINDOW_LENGTH))
    }, [setCpuLoad]
  )

  React.useEffect(
    () => {
      fetchCPU()

      const interval = setInterval(() => {
        fetchCPU()
      }, TEN_SECONDS)

      return () => clearInterval(interval)
    }, [fetchCPU]
  )

  return (
    <CPULoadContext.Provider value={cpuLoad}>
      {children}
    </CPULoadContext.Provider>
  )
}