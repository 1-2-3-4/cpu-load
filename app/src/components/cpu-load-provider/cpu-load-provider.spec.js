import * as React from 'react'
import { shallow } from 'enzyme'

import { CPULoadProvider } from './cpu-load-provider'

describe('CPULoadProvider', () => {
  const child = <div>this is a child</div>
  const mockValue = [10, 20, 30]
  const mockSetValue = jest.fn()
  let useCallbackSpy
  let useEffectSpy
  let useStateSpy
  let wrapper

  beforeAll(() => {
    useCallbackSpy = jest.spyOn(React, 'useCallback')
    useEffectSpy = jest.spyOn(React, 'useEffect').mockImplementation(f => f())
    useStateSpy = jest.spyOn(React, 'useState')
      .mockImplementation(value => [mockValue, mockSetValue])

    wrapper = shallow(
      <CPULoadProvider>
        {child}
      </CPULoadProvider>
    )
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  it('calls useEffect with fetch function', () => {
    expect(useEffectSpy).toHaveBeenCalledWith(expect.any(Function), [expect.any(Function)])
  })

  it('renders children', () => {
    expect(wrapper.contains(child)).toBe(true)
  })

  it('sets provider value to match value from state', () => {
    expect(wrapper.prop('value')).toMatchObject(mockValue)
  })

  it('matches snapshot', () => {
    expect(wrapper.getElement()).toMatchSnapshot()
  })
})