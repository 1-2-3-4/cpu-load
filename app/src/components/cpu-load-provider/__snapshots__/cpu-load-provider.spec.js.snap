// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CPULoadProvider matches snapshot 1`] = `
<Context.Provider
  value={
    Array [
      10,
      20,
      30,
    ]
  }
>
  <div>
    this is a child
  </div>
</Context.Provider>
`;
